static-instrumenter/agent-instrumenter/src/main/java/io/opentelemetry/contrib/statical/instrumenter/ClassArchive.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.contrib.statical.instrumenter;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Enumeration;
+import java.util.Map;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipException;
+import java.util.zip.ZipOutputStream;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Represents an archive storing classes (JAR, WAR). */
+class ClassArchive {
+
+  interface Factory {
+    default ClassArchive createFor(JarFile source, Map<String, byte[]> instrumentedClasses) {
+      return new ClassArchive(source, instrumentedClasses);
+    }
+  }
+
+  private static final Logger logger = LoggerFactory.getLogger(ClassArchive.class);
+
+  private final JarFile source;
+  private final Map<String, byte[]> instrumentedClasses;
+
+  ClassArchive(JarFile source, Map<String, byte[]> instrumentedClasses) {
+    this.source = source;
+    this.instrumentedClasses = instrumentedClasses;
+  }
+
+  void copyAllClassesTo(ZipOutputStream outJar) throws IOException {
+
+    Enumeration<JarEntry> inEntries = source.entries();
+    while (inEntries.hasMoreElements()) {
+      copyEntry(inEntries.nextElement(), outJar);
+    }
+  }
+
+  private void copyEntry(JarEntry inEntry, ZipOutputStream outJar) throws IOException {
+    String inEntryName = inEntry.getName();
+    ZipEntry outEntry =
+        inEntryName.endsWith(".jar") ? new ZipEntry(inEntry) : new ZipEntry(inEntryName);
+
+    try (InputStream entryInputStream = getInputStreamForEntry(inEntry, outEntry)) {
+      try {
+        outJar.putNextEntry(outEntry);
+        StreamUtils.copy(entryInputStream, outJar);
+        outJar.closeEntry();
+      } catch (ZipException e) {
+        if (!isEntryDuplicate(e)) {
+          logger.error("Error while creating entry: " + outEntry.getName(), e);
+          throw e;
+        }
+      }
+    }
+  }
+
+  private static boolean isEntryDuplicate(ZipException ze) {
+    return (ze.getMessage() != null && ze.getMessage().contains("duplicate"));
+  }
+
+  private InputStream getInputStreamForEntry(JarEntry inEntry, ZipEntry outEntry)
+      throws IOException {
+
+    InputStream entryIn = null;
+    ArchiveEntry entry = ArchiveEntry.ofFilePath(inEntry.getName());
+    if (entry.isClass() && !entry.shouldBeSkipped()) {
+      String className = entry.getClassName();
+      try {
+        // sanity check
+        Class.forName(className, false, ClassLoader.getSystemClassLoader());
+        byte[] modified = instrumentedClasses.get(entry.getClassPath());
+
+        if (modified != null) {
+          logger.debug("Found instrumented class: " + className);
+          entryIn = new ByteArrayInputStream(modified);
+          outEntry.setSize(modified.length);
```

Comment:
The size here is only set when instrumented class is encountered - what about other classes, or non-class resources? Is this line needed at all?

Reply:
It will be default (-1 = unknown). I guess it's better to set it if easily available as in this case.